warehouse/query-core/src/test/java/datawave/query/data/parsers/DatawaveKeyTest.java

```diff
@@ -1,88 +1,220 @@
-        Assert.assertEquals("uid", key.getUid());
-        Assert.assertEquals("fieldValueA\u0000blah", key.getFieldValue());
-        Assert.assertEquals("fieldNameA", key.getFieldName());
-        Assert.assertEquals(new Text("row"), key.getRow());
+        assertInvalid(false);
+        assertType(TERM_OFFSETS);
+        assertRow(row);
+        assertDataType(datatype);
+        assertUid(uid);
+        assertFieldName(fieldName);
+        assertFieldValue(fieldValue);
     }
     
     @Test
-    public void parseShardFiKey() {
-        Key testKey = new Key("row", "fi\u0000fieldNameA", "fieldValueA\u0000datatype\u0000uid", "viz");
-        DatawaveKey key = new DatawaveKey(testKey);
+    public void testTermFrequencyKeyWithChildUid() {
+        Key key = new Key(row, "tf", "datatype\u0000uid.1\u0000fieldValue\u0000fieldName", cv);
+        parse(key);
         
-        Assert.assertEquals("datatype", key.getDataType());
-        Assert.assertEquals("uid", key.getUid());
-        Assert.assertEquals("fieldValueA", key.getFieldValue());
-        Assert.assertEquals("fieldNameA", key.getFieldName());
-        Assert.assertEquals(new Text("row"), key.getRow());
+        assertInvalid(false);
+        assertType(TERM_OFFSETS);
+        assertRow(row);
+        assertDataType(datatype);
+        assertUid("uid.1");
+        assertFieldName(fieldName);
+        assertFieldValue(fieldValue);
     }
     
     @Test
-    public void parseShardFiKeyInvalid() {
-        Key testKey = new Key("row", "fi\u0000fieldNameA", "fieldValueA\u0000datatype", "viz");
-        DatawaveKey key = new DatawaveKey(testKey);
-        Assert.assertTrue(key.isInvalidKey());
+    public void testTermFrequencyKeyWithExtraNulls() {
+        Key key = new Key(row, "tf", "datatype\u0000uid\u0000fi\u0000eldVa\u0000lue\u0000fieldName", cv);
+        parse(key);
+        
+        assertInvalid(false);
+        assertType(TERM_OFFSETS);
+        assertRow(row);
+        assertDataType(datatype);
+        assertUid(uid);
+        assertFieldName(fieldName);
+        assertFieldValue("fi\0eldVa\0lue");
     }
     
     @Test
-    public void parseShardTfKey() {
-        Key testKey = new Key("row", "tf", "datatype\u0000uid\u0000fieldValue\u0000fieldName", "viz");
-        DatawaveKey key = new DatawaveKey(testKey);
-        Assert.assertFalse(key.isInvalidKey());
-        Assert.assertEquals("row", key.getShardId());
-        Assert.assertEquals(DatawaveKey.KeyType.TERM_OFFSETS, key.getType());
-        Assert.assertEquals("datatype", key.getDataType());
-        Assert.assertEquals("uid", key.getUid());
-        Assert.assertEquals("fieldValue", key.getFieldValue());
-        Assert.assertEquals("fieldName", key.getFieldName());
+    public void testGlobalIndexKey() {
+        Key key = new Key("fieldValue", "fieldName", "row\0datatype");
+        parse(key);
+        
+        assertInvalid(false);
+        assertType(KeyType.INDEX);
+        assertRow(fieldValue);
```

Comment:
Is the value "row" retrievable from DatawaveKey for global index entries?  It seems odd that both assertRow and assertFieldValue are matching fieldValue.  I know you didn't change the implementation.  Just curious about your thoughts.

Reply:
It is *technically* correct, even if it looks odd. 

It looks like the DatawaveKey was originally a intended to flexibly parse field index, event, and term frequency keys with a single object (all key types from the shard table). However, it also works just as well on global index keys and in that case the field value and row are the same.